File: apps/ui/src/core/clock.ts
/** The current instant, injected so time-dependent rendering is deterministic in
 * tests (e.g. "waiting for 2m"). The frontend converts UTC to local only at render. */
export interface Clock {
  now(): Date;
}

export class SystemClock implements Clock {
  now(): Date {
    return new Date();
  }
}

export class FixedClock implements Clock {
  constructor(private readonly instant: Date) {}

  now(): Date {
    return this.instant;
  }
}

File: packages/py/src/airlock/infrastructure/clock.py
"""Clock implementations."""

from datetime import UTC, datetime


class SystemClock:
    """Production clock: the current instant in UTC."""

    def now(self) -> datetime:
        return datetime.now(UTC)


class FixedClock:
    """A clock pinned to a fixed instant. Makes time deterministic in tests."""

    def __init__(self, instant: datetime) -> None:
        self._instant = instant

    def now(self) -> datetime:
        return self._instant
